### PR TITLE
[IIO] _create() creates the buffer

### PR DESCRIPTION
**Pull request by kparichay**
Create the buffer, allocate memory for the buffer
Enable setting the timestamp for the buffer by base src
Ignore timestamp channel for now - added in todo

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 1]
```diff
@@ -150,6 +151,7 @@ enum
 #define TRIGGER "trigger"
 #define CHANNELS "scan_elements"
 #define IIO "iio:"
+#define TIMESTAMP "timestamp"
```
Comment (wooksong) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: This is a trivial comment. Some string constants have names like ${NAME}_PREFIX  and others have names like just ${NAME}.  Is it intended something with those names?

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: ${NAME} gives name for the element - this element can be device, trigger, buffer, etc
${NAME}_PREFIX is gives prefix for their file/directory name when filtering for the corresponding files/directory.

Please suggest better naming and I will update :)

Comment (wooksong) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: I really have no idea about what are good variable names. I just asked:)

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 2]
```diff
@@ -1603,12 +1612,45 @@ gst_tensor_src_iio_get_times (GstBaseSrc * basesrc, GstBuffer * buffer,
...
+  if (!gst_memory_map (mem, &info, GST_MAP_WRITE)) {
+    goto error_mem_free;
+  }
+  info.size = size;
+  gst_memory_unmap (mem, &info);
+
+  buf = gst_buffer_new ();
+  gst_buffer_append_memory (buf, mem);
+
+  if (gst_tensor_src_iio_fill (src, offset, buffer_size, buf) != GST_FLOW_OK) {
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: [Question]
gst_tensor_src_iio_fill() is vmethod of base-src. Is this called when implementing _create() function?

File: gst/nnstreamer/tensor_source/tensor_src_iio.c [hunk 3]
```diff
@@ -1603,12 +1612,45 @@ gst_tensor_src_iio_get_times (GstBaseSrc * basesrc, GstBuffer * buffer,
...
+  guint buffer_size;
+
+  self = GST_TENSOR_SRC_IIO_CAST (src);
+  buffer_size = gst_tensor_info_get_size (&self->tensors_config->info.info[0]);
+
+  mem = gst_allocator_alloc (NULL, buffer_size, NULL);
+  if (!gst_memory_map (mem, &info, GST_MAP_WRITE)) {
+    goto error_mem_free;
+  }
+  info.size = size;
```
Comment (jaeyun-jung) [inline gst/nnstreamer/tensor_source/tensor_src_iio.c]: In GstMapInfo info; > info.size
Is it necessary? I think we don't need to do gst_memory_map() - gst_memory_unmap()
IMO, just check mem is not null in line 1632.

Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Or check buffer_size == size

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 2 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Yes, implementation of _create() in base-src calls gst_tensor_src_iio_fill().

Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Then I think this is unnecessary, base-src calls_create() and _fill()

Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Please correct me if i am wrong - as per my understanding, as we have overridden the base class _create(), we will have to call _fill() from _create().

Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: I misunderstand your comment. (I understood that basr-src calls both _create() and _fill() , then we don't need to call _fill() in the create() function )
```When implements _create(), then base-src never call the vmethod _fill()``` is it correct?

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 3 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: Added memory check for null
Added bug fix - the function arguments `offset` and `size` should be ignored as they are for pull mode.
`info.size' should be set with buffer_size - and yes, gst_memory_map() - gst_memory_unmap() is reduntant.
updated.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 2 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: > When implements _create(), then base-src never call the vmethod _fill()
Yes. _fill() is only called from vmethod _create() of base-src, which we override.

Comment (jaeyun-jung) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: I understand, thanks for the detailed explanation.

File: gst/nnstreamer/tensor_source/tensor_src_iio.c (hunk 1 above)
Comment (kparichay) [reply gst/nnstreamer/tensor_source/tensor_src_iio.c]: :)

**Pull request merged**